<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>dNdS · NaturalSelection.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../index.html"><img class="logo" src="../assets/logo.png" alt="NaturalSelection.jl logo"/></a><h1>NaturalSelection.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search.html"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../index.html">Home</a></li><li><span class="toctext">Manual</span><ul><li class="current"><a class="toctext" href="dNdS.html">dNdS</a><ul class="internal"><li><a class="toctext" href="#The-NG86-method-1">The NG86 method</a></li><li class="toplevel"><a class="toctext" href="#The-MacDonald-Kreitman-Test-1">The MacDonald Kreitman Test</a></li><li><a class="toctext" href="#References-1">References</a></li></ul></li><li><a class="toctext" href="tajimad.html">Tajima&#39;s D</a></li></ul></li><li><a class="toctext" href="../contributing.html">Contributing</a></li></ul></nav><article id="docs"><header><nav><ul><li>Manual</li><li><a href="dNdS.html">dNdS</a></li></ul><a class="edit-page" href="https://github.com/BioJulia/NaturalSelection.jl/blob/master/docs/src/man/dNdS.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>dNdS</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="dN-/-dS-1" href="#dN-/-dS-1">dN / dS</a></h1><p>NaturalSelection.jl provides several different methods of inferring the action of natural selection from coding sequences.</p><p>Evolutionary pressures on proteins are often quantified by the ratio of substitution rates at non-synonymous and synonymous sites i.e. dN/dS.</p><p>The dN/dS ratio was originally developed for application to distantly diverged sequences, the differences among which represent substitutions that have fixed along independent lineages.</p><p>Nevertheless, the dN/dS measure is often applied to sequences sampled from a single population, the differences among which represent segregating polymorphisms. However, do be careful if this is what you are doing, as it has been demonstrated that dN/dS is not always suitable for such purposes <a href="https://doi.org/10.1371/journal.pgen.1000304">(Sergey Kryazhimskiy &amp; Joshua B. Plotkin, 2008)</a>.</p><h2><a class="nav-anchor" id="The-NG86-method-1" href="#The-NG86-method-1">The NG86 method</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="NaturalSelection.dNdS_NG86" href="#NaturalSelection.dNdS_NG86"><code>NaturalSelection.dNdS_NG86</code></a> — <span class="docstring-category">Function</span>.</div><div><pre><code class="language-none">dNdS_NG86(x, y, addone::Bool = true, code::Int = 1)</code></pre><p>Compute dN and dS, using the <a href="https://www.ncbi.nlm.nih.gov/pubmed/3444411">Nei and Gojobori 1986</a> method.</p><p>The genetic code that is used, is defined according to the numbering of <code>ncbi_trans_table</code>. Code 1 is the standard genetic code.</p><p>This function requires two iterables <code>x</code> and <code>y</code>. If these iterables yield <code>Codon{DNA}</code> or <code>Codon{RNA}</code> type variables. Then it is assumed that <code>x</code> and <code>y</code> are iterables that yield a sequence of aligned codons. If the iterables produce <code>DNA</code> or <code>RNA</code> type variables, then it is assumed <code>x</code> and <code>y</code> iterables that conform to the behaviour of DNA or RNA sequences as defined in the BioSequences package. In this case, a new <code>x</code> and <code>y</code> that do have an element type of <code>Codon{DNA}</code> or <code>Codon{RNA}</code>.</p><p>NG86 is a counting method of computing dN/dS and is typically safer to use on sequence data where codon usage, (esp. at 3rd position), is uniform, the sequences are not very divergent, and transition/transversion rates, are similar.</p></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/NaturalSelection.jl/blob/a6cd4afd5fbd18cddfadf772814746ebafd7470c/src/dnds/NG86/NG86.jl#L8-L27">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="NaturalSelection.S_N_NG86" href="#NaturalSelection.S_N_NG86"><code>NaturalSelection.S_N_NG86</code></a> — <span class="docstring-category">Function</span>.</div><div><pre><code class="language-none">S_N_NG86(codon::C, code::GeneticCode) where {C &lt;: CDN}</code></pre><p>Enumerate the number of synonymous (S) and non-synonymous (N) sites in a codon, using the method used by <a href="https://www.ncbi.nlm.nih.gov/pubmed/3444411">Nei and Gojobori (1986)</a>.</p><p>Returns a tuple where S is the first element and N is the second (S, N).</p><p>Each site in a codon may be both partially synonymous and non-synonymous.</p></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/NaturalSelection.jl/blob/a6cd4afd5fbd18cddfadf772814746ebafd7470c/src/dnds/NG86/computation.jl#L9-L18">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="NaturalSelection.DS_DN_NG86" href="#NaturalSelection.DS_DN_NG86"><code>NaturalSelection.DS_DN_NG86</code></a> — <span class="docstring-category">Function</span>.</div><div><pre><code class="language-none">DS_DN_NG86(x::C, y::C, code::GeneticCode) where C &lt;: CDN</code></pre><p>Compute the number of synonymous (DS) and non-synonymous (DN) mutations between two codons, using the all paths method used by the <a href="https://www.ncbi.nlm.nih.gov/pubmed/3444411">Nei and Gojobori (1986)</a>.</p></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/NaturalSelection.jl/blob/a6cd4afd5fbd18cddfadf772814746ebafd7470c/src/dnds/NG86/computation.jl#L47-L52">source</a></section><h1><a class="nav-anchor" id="The-MacDonald-Kreitman-Test-1" href="#The-MacDonald-Kreitman-Test-1">The MacDonald Kreitman Test</a></h1><p>This test detects and measure the amount of adaptive evolution within a species by determining whether adaptive evolution has occurred, and the proportion of substitutions that resulted from positive selection.</p><p>To do this, the <a href="https://doi.org/10.1038%2F351652a0">McDonald–Kreitman</a> test compares the amount of variation within a species (polymorphism) to the divergence between species (substitutions) at two types of sites, neutral and nonneutral.</p><p>A substitution refers to a nucleotide that is fixed within one species, but a different nucleotide is fixed within a second species at the same base pair of homologous DNA sequences.</p><p>The two types of sites can be either synonymous or nonsynonymous within a protein-coding region.</p><p>The null hypothesis of the McDonald–Kreitman test is that the ratio of nonsynonymous to synonymous variation within a species is going to equal the ratio of nonsynonymous to synonymous variation between species (i.e. Dn/Ds = Pn/Ps).</p><p>When positive or negative selection (natural selection) influences nonsynonymous variation, the ratios will no longer equal. The ratio of nonsynonymous to synonymous variation between species is going to be lower than the ratio of nonsynonymous to synonymous variation within species (i.e. Dn/Ds &lt; Pn/Ps) when negative selection is at work, and deleterious mutations strongly affect polymorphism. The ratio of nonsynonymous to synonymous variation within species is lower than the ratio of nonsynonymous to synonymous variation between species (i.e. Dn/Ds &gt; Pn/Ps) when we observe positive selection.</p><p>Under neutrality the expectation is (Pn / Ps) == (Dn / Ds).</p><p>The statistic <a href="https://doi.org/10.1038%2F4151022a">α</a> represents the proportion of substitutions driven by positive selection. α can be equal to any number between -Inf and 1. Negative values of alpha are produced by sampling error, assumption violations, or the segregation of <a href="https://doi.org/10.1093/molbev/msn005">slightly deleterious</a> amino acid mutations. The null hypothesis here is that α = 0.</p><p>The neutrality index (NI) quantifies the direction and degree of departure from neutrality (where Pn/Ps and Dn/Ds ratios equal). A neutrality index greater than 1 is indicative of negative selection, a neutrality index lower than 1 indicates positive selection is at work in the population.</p><p>This test is provided by the function <code>mkt</code>.</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="NaturalSelection.mkt" href="#NaturalSelection.mkt"><code>NaturalSelection.mkt</code></a> — <span class="docstring-category">Function</span>.</div><div><pre><code class="language-none">mkt(x, y, ref)</code></pre><p>Compute <a href="https://doi.org/10.1038%2F351652a0">McDonald-Kreitman test</a> statistics for two sets of codon sequences.</p><p>This function returns the values of Ps, Pn, Ds, Dn, α, and the neutrality index (NI).</p></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/NaturalSelection.jl/blob/a6cd4afd5fbd18cddfadf772814746ebafd7470c/src/dnds/mkt/mkt.jl#L71-L78">source</a></section><h2><a class="nav-anchor" id="References-1" href="#References-1">References</a></h2><p>For more about the McDonald Kreitman Test, see the following references:</p><p>https://en.wikipedia.org/wiki/McDonald%E2%80%93Kreitman_test</p><p><a href="https://doi.org/10.1038%2F351652a0">McDonald, J. H. Kreitman (1991)</a></p><p><a href="https://doi.org/10.1093%2Fmolbev%2Fmsn005">Charlesworth, J. Eyre-Walker (2008)</a></p><p><a href="http://www.zoology.wisc.edu/courses/611/Part2/Readings/3eyre-walker_tree2006.pdf">Eyre-Walker, A (2006)</a></p><p><a href="https://doi.org/10.1093%2Fmolbev%2Fmsq249">Stoletzki, N. Eyre-Walker (2010)</a></p><footer><hr/><a class="previous" href="../index.html"><span class="direction">Previous</span><span class="title">Home</span></a><a class="next" href="tajimad.html"><span class="direction">Next</span><span class="title">Tajima&#39;s D</span></a></footer></article></body></html>
